### ISSUE TITLE
Extend restraints to explicit solvent case

### ISSUE THREAD

**Issue by davidlmobley** [choderalab/yank#245]
Currently, restraints only are implemented for the implicit solvent case. They ought to be extended to handle the explicit solvent case. Are there any plans to do this? 

As I understand it from @nividic in my group, the challenge is that the custom bonded force currently used for restraints will only work in nonperiodic systems, so the implementation for explicit solvent with cutoffs will need to use a different mechanism (custom nonbonded force, perhaps). 

In my view, clearly this needs to be done, because as far as I know:
- No one has looked at the relative efficiency of running with restraints vs running without restraints
- The (old) data I have from ~2006 suggests that restraints will make some aspects substantially more efficient
- Conventionally we've argued (and I still think) that using restraints avoids problems with ligand stickiness at odd points in the protein at intermediate lambda values which slow convergence

Have you guys looked at these issues in implicit solvent and concluded that restraints are not needed in explicit solvent for some reason? I tried searching on here for info along those lines but didn't find anything. 

If this hasn't already been checked, it seems like the two next logical steps would be:
a) In implicit solvent, compare efficiency of running with restraints versus running without restraints
b) Implement restraints for explicit solvent (@nividic says he thinks he knows how it could be done, so we could probably get involved with this if necessary). 

I'll also raise a separate issue for another aspect I'd like to discuss relating to restraints.

**Comment by jchodera** [comment]
We've decided on a course of action where I make it easy to add new `Restraint` subclasses that specify their own command-line options and are autodetected by the command-line interface.  This should be quick to add, but we're still dealing with some bugs at the moment.

**Comment by Lnaden** [comment]
Was this fixed by #500 or do we still need the more generalized fixes in before we can resolve this?

**Comment by andrrizzi** [comment]
Not yet! Currently restraints are still added only to implicit solvent.
